Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+        } else {
+            return verticalItemsCount
+        }
+    }
+
+    func maxHorizontalItemsCount() -> Int {
+        let horizontalItemsCount =  Int(floor(boundsSize.width / (itemSize.width + insets.left)))
+        if let delegate = self.collectionView?.delegate as? ASHorizontalScrollDelegate {
+            return delegate.numberOfHorizontalItems(horizontalItemsCount)
+        } else {
+            return horizontalItemsCount
+        }
+    }
+    func collectionViewSizeForRect(contentSize: CGSize) -> CGSize {
+        let horizontalItemsCount = maxHorizontalItemsCount()
+        let verticalItemsCount = maxVerticalItemsCount()
+
+
+        // Take the number of cells and subtract its space in the view from the height. The left over space is the white space.
+        // The left over space is then devided evenly into (n + 1) parts to figure out how much space should be inbetween a cell
+        var verticalInsets = (contentSize.height - (CGFloat(verticalItemsCount) * itemSize.height)) / CGFloat(verticalItemsCount + 1)
+        var horizontalInsets = (contentSize.width - (CGFloat(horizontalItemsCount) * itemSize.width)) / CGFloat(horizontalItemsCount + 1)
+
+        // We want a minimum inset to make things not look crowded. We also don't want uneven spacing.
+        // If we dont have this. Set a minimum inset and recalculate the size of a cell
+        if horizontalInsets < minimumInsets || horizontalInsets != verticalInsets {
+            verticalInsets = minimumInsets
+            horizontalInsets = minimumInsets
+            itemSize.width = (contentSize.width - (CGFloat(horizontalItemsCount + 1) * horizontalInsets)) / CGFloat(horizontalItemsCount)
+            itemSize.height = itemSize.width
+        }
+
+        let itemsPerPage = verticalItemsCount * horizontalItemsCount
+
+        numberOfPages = Int(ceil(Double(cellCount) / Double(itemsPerPage)))
+
+
+        insets = UIEdgeInsets(top: verticalInsets, left: horizontalInsets, bottom: verticalInsets, right: horizontalInsets)
+
+        var size = contentSize
+        size.width = CGFloat(numberOfPages) * contentSize.width
+
+        return size
+    }
+
+    override func layoutAttributesForElementsInRect(rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        var allAttributes = [UICollectionViewLayoutAttributes]()
+        for i in 0 ..< cellCount {
+            let indexPath = NSIndexPath(forRow: i, inSection: 0)
+            let attr = self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+            allAttributes.append(attr)
+        }
+        return allAttributes
+    }
+
+    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
+        return self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+    }
+
+    override func shouldInvalidateLayoutForBoundsChange(newBounds: CGRect) -> Bool {
+        if newBounds.width < 320 {
```

Comment:
When the screen width is 320 I do want to invalidate layout. 
Sometimes the bounds of the collectionview becomes 100,75 or something very tiny. It usually happens when the orientation changes with the topsites not on screen. In those cases I dont want to invlidate the existing layout. I'll add a comment to explain what is going on here.